crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+    ///    +------------+----------------+------------------------------------+
+    /// ```
+    pub fn emit_tsig_for_mac(
+        &self,
+        encoder: &mut BinEncoder<'_>,
+        key_name: &Name,
+    ) -> ProtoResult<()> {
+        key_name.emit_as_canonical(encoder, true)?;
+        DNSClass::ANY.emit(encoder)?;
+        encoder.emit_u32(0)?; // TTL
+        self.algorithm.emit(encoder)?;
+        encoder.emit_u16((self.time >> 32) as u16)?;
+        encoder.emit_u32(self.time as u32)?;
+        encoder.emit_u16(self.fudge)?;
+        encoder.emit_u16(self.error)?;
+        encoder.emit_u16(self.other.len() as u16)?;
+        encoder.emit_vec(&self.other)?;
+        Ok(())
+    }
+
+    /// Add actual MAC value to existing TSIG record data.
+    ///
+    /// # Arguments
+    ///
+    /// * `mac` - mac to be stored in this record.
+    pub fn set_mac(self, mac: Vec<u8>) -> Self {
+        assert!(mac.len() < (1 << 16));
+        TSIG { mac, ..self }
+    }
+}
+
+/// Read the RData from the given Decoder
+///
+/// ```text
+///                       1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  /                         Algorithm Name                        /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                                                               |
+///  |          Time Signed          +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                               |            Fudge              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          MAC Size             |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+             MAC               /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Original ID          |            Error              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Other Len            |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn read(decoder: &mut BinDecoder<'_>) -> ProtoResult<TSIG> {
+    let algorithm = Algorithm::read(decoder)?;
+    let time_high = decoder.read_u16()?.unverified(/*valid as any u16*/) as u64;
+    let time_low = decoder.read_u32()?.unverified(/*valid as any u32*/) as u64;
+    let time = (time_high << 32) + time_low;
+    let fudge = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let mac_size = decoder.read_u16()?.unverified(/* TODO maybe compare to alg out size, but we don't actually know that value? */);
```

Comment:
In regards the concern here is if the foreign data is trusted... so `mac_size` should be validated to be less than the remainder of the `Record` being read. Here's an example: https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/rr/dnssec/rdata/sig.rs#L469-L473

Otherwise the issue is that a malicious actor could cause this to read past the end of the record during deserialization by passing in a bad length and cause a crash or other badness. That other code is a little older, I think you can just use `len < decoder.len()`, as that will ensure that it's less than what is left in the decoder.